android/app/src/main/java/app/organicmaps/bookmarks/BookmarkListAdapter.java

```diff
@@ -96,12 +96,13 @@ private void calculateSections()
       mBookmarksSectionIndex = SectionPosition.INVALID_POSITION;
       mTracksSectionIndex = SectionPosition.INVALID_POSITION;
 
+      BookmarkCategory category = BookmarkManager.INSTANCE.getCategoryById(getCategory().getId());
       mSectionsCount = 0;
       // We must always show the description, even if it's blank.
       mDescriptionSectionIndex = mSectionsCount++;
-      if (getCategory().getTracksCount() > 0)
```

Comment:
Why getCategory returns not updated value? Should its cache be somehow refreshed/updated after a track was deleted/moved?

Reply:
Actually the onChange method is called when adapter.notifyDataSetChanged() method is called, which retrieves the data from core and replaces it with current cached data.

But this updateSectionIndexes() method is called before that which calculates section indexes. Since data gets updated after this method so section indexes gets wrong if any of the intermediate section gets removed.

Technically section indexes should be calculated after updation of the data.

Reply:
So what would be the best/cleanest solution here?

@Lampotrias do you have any ideas here?

P.S. We are not afraid of rewriting/refactoring code in Organic Maps if necessary.

Reply:
> So what would be the best/cleanest solution here?
> 
> 

Problem is only there when list is not sorted (i.e. sorting mode is default), if list is sorted then everything works fine because in that case before recalculation the deleted object is removed from the local sorted list (independent of core). But in case of default we just need counts of tracks and bookmarks, and its done without doing any changes locally (probably it assumes that bookmark category data has been updated when changes were done in core)

Opt 1 (current solution) - we just need track and bookmark count so we can fetch it directly from core while calculation

Opt 2 - During deletion we can decrease the count locally as well  to keep the count synced.

Opt 3 - We can somehow manage to trigger the onChange() method which replaces the old category data, after doing any core related operation (like track deletion, category change etc.)

Reply:
> So what would be the best/cleanest solution here?
> 
> @Lampotrias do you have any ideas here?
> 
> P.S. We are not afraid of rewriting/refactoring code in Organic Maps if necessary.

The solution to this problem isn't straightforward, and @kavikhalique was right—there's a race condition here. I'll try to resolve it with minimal changes to the code. I believe this issue doesn't affect https://github.com/organicmaps/organicmaps/pull/9033.

Reply:
> Opt 1 (current solution) - we just need track and bookmark count so we can fetch it directly from core while calculation
> 
> Opt 2 - During deletion we can decrease the count locally as well to keep the count synced.
> 
> Opt 3 - We can somehow manage to trigger the onChange() method which replaces the old category data, after doing any core related operation (like track deletion, category change etc.)

Opt 1 can be left as a temporary solution for now. It’s simple and solves the problem without causing any errors.

The current solution is fundamentally wrong, and it’s just lucky that it works when onChange() is triggered by any update in the Adapter. We should first update the dataset and only then call notifyDataSetChanged() to render the data on the screen. Right now, it's the opposite—we call notifyDataSetChanged(), some rendering happens, onChange() is called, and only then do we change the data, after the rendering is already done.

If opt 1 isn’t suitable, I’ll look for a solution from “opt 2.”

P.S. @kavikhalique, did you actually get an app crash on my PR? Did the stack trace not get saved?

Reply:
> The current solution is fundamentally wrong, and it’s just lucky that it works when onChange() is triggered by any update in the Adapter. We should first update the dataset and only then call notifyDataSetChanged() to render the data on the screen. Right now, it's the opposite—we call notifyDataSetChanged(), some rendering happens, onChange() is called, and only then do we change the data, after the rendering is already done.
> 

Yes, i think while designing it was assumed that whenever notifyDatasetChanged() will be called it will first trigger the onChange() method and after that it will reconstruct the list but i think it is happening opposite.


> If opt 1 isn’t suitable, I’ll look for a solution from “opt 2.”
> 

In this case i would suggest initially go with the approach which u suggested above. Try to find a way that whenever notifyDataSetChanged() method is called then onChange() method get triggered before list update. This will require less changes and more compatibility. 

If there is no way to do that then second better way would be to create local variables for track count and bookmark count similar to sorted section datasource, and changing those values whenever track count or bookmark count changes (try to cover all cases in which track/bookmark count changes) and update those local variables accordingly.

The third solution would be as u did before i.e. updating datasource at the time of trigger. (Similar to second one here also you have to look for all places where track and bookmark count is being changed). I am not very sure about semantics of this but if it works and above two fails then its fine.

If you have any better design we can discuss that as well.

> 
> P.S. @kavikhalique, did you actually get an app crash on my PR? Did the stack trace not get saved?

Yes, it crashed. I have provided the steps to regenerate it please have a look at it.

Reply:
> Yes, it crashed. I have provided the steps to regenerate it please have a look at it.
No need, I have a rough understanding of the cause.

I’ll try to find a more stable solution.

Reply:
https://github.com/organicmaps/organicmaps/pull/9257